flink-streaming-connectors/flink-connector-elasticsearch2/src/main/java/org/apache/flink/streaming/connectors/elasticsearch2/ElasticsearchSink.java

```diff
@@ -186,22 +191,44 @@ public void beforeBulk(long executionId, BulkRequest request) {
 
 			@Override
 			public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+				boolean allRequestsRepeatable = true;
 				if (response.hasFailures()) {
 					for (BulkItemResponse itemResp : response.getItems()) {
 						if (itemResp.isFailed()) {
-							LOG.error("Failed to index document in Elasticsearch: " + itemResp.getFailureMessage());
-							failureThrowable.compareAndSet(null, new RuntimeException(itemResp.getFailureMessage()));
+							// Check if index request can be retried
+							String failureMessageLowercase = itemResp.getFailureMessage().toLowerCase();
+							if (failureMessageLowercase.contains("timeout") || failureMessageLowercase.contains("timed out") // Generic timeout errors
+									|| failureMessageLowercase.contains("UnavailableShardsException".toLowerCase()) // Shard not available due to rebalancing or node down
+									|| (failureMessageLowercase.contains("data/write/bulk") && failureMessageLowercase.contains("bulk")) // Bulk index queue on node full 
+								) {
+								LOG.debug("Retry batch: " + itemResp.getFailureMessage());
```

Comment:
I can not assess how often retries are needed.
Users can also manually increase the log level if needed. So we can leave it as is.
However, I'm wondering whether we want to include a metric that counts the number of retries that occurred.